### DISCUSSION TITLE
Override alternation query

### DISCUSSION THREAD

**Discussion by syyyr** [nvim-treesitter/nvim-treesitter#6089] [Q&A] [answered]
Hi,

in the query file for c, there's this query:
```
[
  (type_qualifier)
  (gnu_asm_qualifier)
  "__extension__"
] @type.qualifier
```
This works fine, however, it also highlights `constexpr` as `@type.qualifier` which is not what I want (I just want it to be `@keyword`). I tried putting this:
```query
; extends

((type_qualifier) @type.qualifier (#not-lua-match? @type.qualifier "constexpr"))
```
I put this in `.config/nvim/queries/c/highlights.scm`. This unfortunately doesn't do what I need. How do I keep the default queries and disable highlighting `constexpr` as a type qualifier?

**Comment by clason** [comment]
You cannot _remove_ or override individual patterns. Instead, add a pattern for `constexpr` (using `#eq?`, not a pointless and expensive literal pattern) that assigns a different capture.

**Reply by syyyr** [reply]
That was what I expected my example to do. How do I add a pattern that assigns a different capture?

**Reply by syyyr** [reply]
Like this?
```
; extends

((type_qualifier) @keyword (#eq? @keyword "constexpr"))
```
This makes constexpr have this kind of highlighting:
```
Treesitter                                                                                                                                                                                     
  - @type.qualifier.cpp links to cStorageClass cpp                                                                                                                                             
  - @keyword.cpp links to Keyword cpp                                                                                                                                                          
  - @keyword.cpp links to Keyword cpp
```
which unfortunately doesn't change anything

**Reply by clason** [reply]
That looks correct. Make sure nothing is interfering with runtimepath ordering; otherwise add a higher priority to your new pattern.

**Reply by syyyr** [reply]
I'm not sure if my runtimepath is correct. This is the output of `:echo nvim_get_runtime_file('queries/c/*.scm', v:true)`
```
['/home/vk/.config/nvim/queries/c/highlights.scm', '/home/vk/.local/share/nvim/treesitter/queries/c/folds.scm', '/home/vk/.local/share/nvim/treesitter/queries/c/highlights.scm', '/home/vk/.local/share/nvim/treesitter/queries/c/indents.scm', '/home/vk/.local/share/nvim/treesitter/queries/c/injections.scm', '/home/vk/.local/share/nvim/treesitter/queries/c/locals.scm', '/home/vk/.loca
l/share/nvim/site/queries/c/folds.scm', '/home/vk/.local/share/nvim/site/queries/c/highlights.scm', '/home/vk/.local/share/nvim/site/queries/c/indents.scm', '/home/vk/.local/share/nvim/site/queries/c/injections.scm', '/home/vk/.local/share/nvim/site/queries/c/locals.scm', '/usr/share/nvim/runtime/queries/c/folds.scm', '/usr/share/nvim/runtime/queries/c/highlights.scm', '/usr/share/
nvim/runtime/queries/c/injections.scm']
```

How do I add priority to the query? It seems to me that it should be done like this:
```
((type_qualifier) @keyword (#eq? @keyword "constexpr") (#set! "priority" 200))
```
but this makes no difference, the output of `:Inspect` is still the same.

**Reply by clason** [reply]
Yes, but the highlight should be different.

**Reply by syyyr** [reply]
Sorry, I should've explained my situation better. The highlighting for constexpr is truly different. My current situation looks like this:

`:Inspect` on `const` returns this:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/17759291/e8063cf6-a1ce-43ab-9350-2419f4acf199)
and `:Inspect` on `constexpr` returns this:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/17759291/c701b30d-3062-46e8-88ee-faa1a83ac64c)

So, `constexpr` gets correctly recognized as `@keyword`. However, my problem is that it combines highlighting from both `cStorageClass` [1], and `@keyword`. Specifically, it gets `@keyword`'s color and `cStorageClass`' boldness. I would like `constexpr` to be highlight without the boldness as is `static` in this screenshot:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/17759291/5bee2e20-9b02-4ed4-a493-2404f21c3cf6)


[1] I've set up the linking to cStorageClass myself via `highlight! link @type.qualifier.cpp cStorageClass`

**Reply by clason** [reply, marked as answer]
Yes, that's a very different issue. Don't do that, then; define the full set of colors and attributes for each highlight group.

**Reply by syyyr** [reply]
Thank you very much. Sorry for not immediately sending the screenshots. In the end, I came up with this:
```
; extends

"constexpr" @keyword.constexpr
```
in `.config/nvim/queries/cpp/highlights.scm` and then
```vim
highlight @keyword.constexpr cterm=nocombine gui=nocombine
```
which results in `constexpr` showing this in `:Inspect`:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/17759291/35432b3a-32a8-4175-a1c2-0b30762de22b)
And the result highlighting looks like this, which is what I want:
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/17759291/ac3a840c-d8ca-449a-9255-ad75fc7d5a5a)

Thanks again for being patient with me.
